pkg/proc/riscv64_arch.go

```diff
@@ -0,0 +1,365 @@
+		// When there's no frame descriptor entry use BP (the frame pointer) instead
+		// - return register is [bp + a.PtrSize()] (i.e. [cfa-a.PtrSize()])
+		// - cfa is bp + a.PtrSize()*2
+		// - bp is [bp] (i.e. [cfa-a.PtrSize()*2])
+		// - sp is cfa
+
+		// When the signal handler runs it will move the execution to the signal
+		// handling stack (installed using the sigaltstack system call).
+		// This isn't a proper stack switch: the pointer to g in TLS will still
+		// refer to whatever g was executing on that thread before the signal was
+		// received.
+		// Since go did not execute a stack switch the previous value of sp, pc
+		// and bp is not saved inside g.sched, as it normally would.
+		// The only way to recover is to either read sp/pc from the signal context
+		// parameter (the ucontext_t* parameter) or to unconditionally follow the
+		// frame pointer when we get to runtime.sigreturn (which is what we do
+		// here).
+
+		return &frame.FrameContext{
+			RetAddrReg: regnum.RISCV64_PC,
+			Regs: map[uint64]frame.DWRule{
+				regnum.RISCV64_PC: {
+					Rule:   frame.RuleOffset,
+					Offset: int64(-a.PtrSize()),
+				},
+
+				regnum.RISCV64_FP: {
+					Rule:   frame.RuleOffset,
+					Offset: int64(-2 * a.PtrSize()),
+				},
+
+				regnum.RISCV64_SP: {
+					Rule:   frame.RuleValOffset,
+					Offset: 0,
+				},
+			},
+
+			CFA: frame.DWRule{
+				Rule:   frame.RuleCFA,
+				Reg:    regnum.RISCV64_FP,
+				Offset: int64(2 * a.PtrSize()),
+			},
+		}
+	}
+
+	if a.crosscall2fn == nil {
+		a.crosscall2fn = bi.lookupOneFunc("crosscall2")
+	}
+
+	if a.crosscall2fn != nil && pc >= a.crosscall2fn.Entry && pc < a.crosscall2fn.End {
+		rule := fctxt.CFA
+
+		if rule.Offset == crosscall2SPOffsetBad {
+			rule.Offset += crosscall2SPOffset
+		}
+		fctxt.CFA = rule
+	}
+
+	// We assume that FP is the frame pointer and we want to keep it updated,
+	// so that we can use it to unwind the stack even when we encounter frames
+	// without descriptor entries.
+	// If there isn't a rule already we emit one.
+	if fctxt.Regs[regnum.RISCV64_FP].Rule == frame.RuleUndefined {
+		fctxt.Regs[regnum.RISCV64_FP] = frame.DWRule{
+			Rule:   frame.RuleFramePointer,
+			Reg:    regnum.RISCV64_FP,
+			Offset: 0,
+		}
+	}
+
+	if fctxt.Regs[regnum.RISCV64_LR].Rule == frame.RuleUndefined {
+		fctxt.Regs[regnum.RISCV64_LR] = frame.DWRule{
+			Rule:   frame.RuleRegister,
+			Reg:    regnum.RISCV64_LR,
+			Offset: 0,
+		}
+	}
+
+	return fctxt
+}
+
+const riscv64cgocallSPOffsetSaveSlot = 0x8
+const riscv64prevG0schedSPOffsetSaveSlot = 0x10
+
+func riscv64SwitchStack(it *stackIterator, callFrameRegs *op.DwarfRegisters) bool {
+	if it.frame.Current.Fn == nil {
+		if it.systemstack && it.g != nil && it.top {
+			it.switchToGoroutineStack()
+			return true
+		}
+		return false
+	}
+	switch it.frame.Current.Fn.Name {
+	case "runtime.cgocallback_gofunc", "runtime.cgocallback":
```

Comment:
This stuff doesn't work or the cgo stacktrace tests would work, delete it.

Reply:
Deleted.

Reply:
I meant all of it: runtime.asmcgocall and crosscall2 are the same thing as is the switch at the bottom of this function.

Reply:
`runtime.asmcgocall` and `crosscall2` removed.